Add Maxpool to shape analysis / Opinfo

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64323 Add embedding shape analysis
* #64121 Max Pool with indices
* **#63530 Add Maxpool to shape analysis / Opinfo**

how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Differential Revision: [D30738147](https://our.internmc.facebook.com/intern/diff/D30738147)

Commit:
Add Maxpool to shape analysis

[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

AnirudhDagar commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2272,6 +2272,29 @@ def generator():
+        for kernel, stride, N, C, H, W, ceil_mode, padding, dilation in product(kerneli, stridei, Ni, Ci, Hi, Wi, ceil_modei, paddingi, dilationi):
+            max_pool = torch.nn.MaxPool2d(kernel, stride, ceil_mode=ceil_mode, padding=padding, dilation=dilation)
+
+            yield SampleInput(make_arg((N, C, H, W)), args=(max_pool.kernel_size, max_pool.stride,
```
Although the ordering for positional args is correct here, I don't think it's a good idea to pass all of the params as positional arguments as mentioned #62545.

eellison replied:
do you mean as keyword arguments ?

AnirudhDagar replied:
Yeah I mean it would be better to add keyword arguments since positional args ordering is not correct.

eellison replied:
Good suggestion! Done

AnirudhDagar commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2272,6 +2272,29 @@ def generator():
+
+    kerneli = [[3, 2], [3, 3]]
+    stridei = [[2, 2]]
+    Ni = [8, 16]
```
Nit: Would be nice to decrease the batch size here, making the test slightly faster I guess.

AnirudhDagar reviewed (commented):
@eellison I'm sure you have a better idea, but I've made some comments for you to take a look at. Also, thanks for the shape analysis addition.

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"




[ghstack-poisoned]

zou3519 reviewed (commented):
The OpInfo parts seem reasonable to me. Could we add a test for `return_indices=True`?

zou3519 reviewed (commented):
shape checking code looks correct, modulo one line. This also caused me to identify some more cases for max_pool we should test

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

zou3519 commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -133,6 +133,72 @@ const std::string shape_compute_functions =
+        def max_pool2d(input: List[int], kernel_size: List[int], stride: List[int], padding: List[int], dilation: List[int], ceil_mode: bool):
+          assert len(kernel_size) == 1 or len(kernel_size) == 2, "max_pool2d: kernel_size must either be a single int, or a tuple of two ints"
+          kH = kernel_size[0]
+          kW = 1 if len(kernel_size) == 1 else kernel_size[1]
```
Speaking of negative tests (https://github.com/pytorch/pytorch/issues/63766)... here's a use case @mruberry

zou3519 replied:
(but yeah, no need to test this in this PR)

zou3519 commented:
>  by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Have you brought this up with @pytorch/pytorch-dev-infra team or more generally? This seems like a really cool code browser

eellison commented:
@zou3519 I have not! Maybe we should reference it on pytorch docs ?

eellison commented:
@zou3519 - i think all the comments have been addressed except return_indices=True which I've stacked.. accept ?

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

seemethere commented:
> > by going to [codebrowser.bddppq.com/pytorch/pytorch](https://codebrowser.bddppq.com/pytorch/pytorch/) typing in native::{op_name} and looking at the op implementation as a bonus if you want
> 
> Have you brought this up with @pytorch/pytorch-dev-infra team or more generally? This seems like a really cool code browser

Oh wow this is really cool!

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want


[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738147).

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Differential Revision: [D30738147](https://our.internmc.facebook.com/intern/diff/D30738147)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738147).

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Differential Revision: [D30738147](https://our.internmc.facebook.com/intern/diff/D30738147)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738147).

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Differential Revision: [D30738147](https://our.internmc.facebook.com/intern/diff/D30738147)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738147).

Commit:
Update on "Add Maxpool to shape analysis / Opinfo"


how to review: pretty much just check that the inputs generated are a good representation of the op semantics, that should be sufficient for correctness, and then you can also double check the op size semantics by going to https://codebrowser.bddppq.com/pytorch/pytorch/ typing in native::{op_name} and looking at the op implementation as a bonus if you want

Differential Revision: [D30738147](https://our.internmc.facebook.com/intern/diff/D30738147)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738147).

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2560,6 +2560,37 @@ def generator():
+    Ni = [1, 4, None]
+    Ci = [32]
+    Hi = [8, 16]
+    Wi = [8, 16]
```
this breaks old gradcheck https://app.circleci.com/pipelines/github/pytorch/pytorch/382098/workflows/b671f508-8d28-43e9-a31f-36f17dd72e06/jobs/16090317, biggest sample (4x32x16x16) requires 2GB allocation for old full jacobian check (ninputs x noutputs in double precision comes to 2147483648 bytes, give or take).
Samples for OpInfos should be smaller, given the gradchecks that are run.

albanD replied:
Sending a PR to fix that by reducing all the sizes significantly.

eellison replied:
sorry about that ! i think i copied these inputs from another test but those tests must not have been doing gradgrad..
